### PR TITLE
fix(sql): fix sample by first/last keyed by non-symbol column

### LINKED ISSUE #2636: incorrect `FILL` results
### Describe the bug

```
select
  id,
  time,
  last(lat) lat,
  last(lon) lon,
  geo6
from
  pos
where
  id = 'YWPCEGICTJSGGBRIJCQVLJ' sample by 15m fill(null) ALIGN to CALENDAR;
```

### To reproduce

compare SQL output with and without `fill` clause

### Expected Behavior

_No response_

### Environment

```markdown
Demo
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by bziobrowski**
Fixes #2636 

PR improves logic of sample by factory selection .
It doesn't allow using SampleByFirstLastRecordCursorFactory when column list contains anything other than  : 
- first/last aggregates
- designated timestamp column
- filtered symbol column

It also enables order by geohash columns . 

In master branch SQL such as : 

```sql
CREATE TABLE pos (
  id int
  geo GEOHASH(5b)
);
insert into pos  values ( 1,##00001), ( 2,##00010), ( 3, ##00011 ), ( 16, ##10000 );
select * from pos order by geo desc;
```
triggers:

`
java.lang.UnsupportedOperationException
        at io.questdb.griffin.engine.functions.AbstractGeoHashFunction.getSymbolB(AbstractGeoHashFunction.java:154)
        at io.questdb.cairo.sql.VirtualRecord.getSymB(VirtualRecord.java:198)
        at io.questdb.griffin.engine.orderby.LongTreeChain.put(LongTreeChain.java:75)
        at io.questdb.griffin.engine.orderby.SortedLightRecordCursor.buildChain(SortedLightRecordCursor.java:130)
        at io.questdb.griffin.engine.orderby.SortedLightRecordCursor.hasNext(SortedLightRecordCursor.java:79)
        at io.questdb.cutlass.http.processors.JsonQueryProcessorState.onQuerySetupFirstRecord(JsonQueryProcessorState.java:755)
        at io.questdb.cutlass.http.processors.JsonQueryProcessorState.doFirstRecordLoop(JsonQueryProcessorState.java:463)
        at io.questdb.cutlass.http.processors.JsonQueryProcessorState.onQueryPrefix(JsonQueryProcessorState.java:719)
        at io.questdb.cutlass.http.processors.JsonQueryProcessorState.resume(JsonQueryProcessorState.java:864)
        at io.questdb.cutlass.http.processors.JsonQueryProcessor.doResumeSend(JsonQueryProcessor.java:306)
        at io.questdb.cutlass.http.processors.JsonQueryProcessor.executeSelect(JsonQueryProcessor.java:550)
        at io.questdb.cutlass.http.processors.JsonQueryProcessor.executeNewSelect(JsonQueryProcessor.java:534)
        at io.questdb.cutlass.http.processors.JsonQueryProcessor.compileQuery(JsonQueryProcessor.java:423)
        at io.questdb.cutlass.http.processors.JsonQueryProcessor.execute0(JsonQueryProcessor.java:186)
        at io.questdb.cutlass.http.processors.JsonQueryProcessor.onRequestComplete(JsonQueryProcessor.java:236)
        at io.questdb.cutlass.http.HttpConnectionContext.handleClientRecv(HttpConnectionContext.java:600)
        at io.questdb.cutlass.http.HttpConnectionContext.handleClientOperation(HttpConnectionContext.java:203)
        at io.questdb.cutlass.http.HttpServer$1.lambda$$0(HttpServer.java:86)
        at io.questdb.network.AbstractIODispatcher.processIOQueue(AbstractIODispatcher.java:178)
        at io.questdb.cutlass.http.HttpServer$1.run(HttpServer.java:101)
        at io.questdb.mp.Worker.run(Worker.java:118)
`

because comparator uses  default (symbol) type .

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 47 / 48 (97.92%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|31|32|96.88%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/RecordComparatorCompiler.java|16|16|100.00%|

**Review by ideoma — DISMISSED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 47 / 48 (97.92%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|31|32|96.88%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/RecordComparatorCompiler.java|16|16|100.00%|

File: core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java [hunk 1]
```diff
@@ -210,11 +210,34 @@ public RecordCursorFactory generateExplain(ExplainModel model, SqlExecutionConte
...
+
         for (int i = 0, n = columns.size(); i < n; i++) {
             final QueryColumn column = columns.getQuick(i);
             final ExpressionNode node = column.getAst();
 
-            if (node.type != ExpressionNode.LITERAL) {
+            if (node.type == LITERAL) {
+                int idx = metadata.getColumnIndex(node.token);
+                if (idx < 0) {
+                    return false;
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java:223]: this line is new and untested

**Review by bluestreak01 — CHANGES_REQUESTED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 46 / 46 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/RecordComparatorCompiler.java|16|16|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|30|30|100.00%|

File: core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java (hunk 1 above)
Comment (bziobrowski) [reply core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java]: Removed in latest commit .

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 46 / 46 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/RecordComparatorCompiler.java|16|16|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|30|30|100.00%|

**Review by bluestreak01 — APPROVED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 46 / 46 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/RecordComparatorCompiler.java|16|16|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|30|30|100.00%|

**Pull request merged**